File: scrape.mjs
import axios from 'axios';
import * as cheerio from 'cheerio';
import { response } from 'express';
import fs, { copyFileSync } from 'fs';
import path from 'path';
import url from 'url';

const __dirname = path.dirname(url.fileURLToPath(import.meta.url));

const BASE_URL = 'https://iansvivarium.com/morphs';
const TYPE_MAP = {
  'Wildtypes': 0,
  'Single Recessive': 1,
  'Double Trait': 2,
  'Triple Trait': 3,
  'Quad Trait': 4,
  'Five Trait': 5,
  'Six Trait': 6,
  'Single Dominant': 7,
  'Single Incomplete Dominant': 7,
  'Selectively Bred': 8,
};

function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

function clean(href) {
    const url = new URL(href, BASE_URL);
    const searchParams = new URLSearchParams(url.search);
    searchParams.delete('sid');
    url.search = searchParams.toString();
    const updatedHref = url.pathname + url.search;
    return updatedHref;
}

async function downloadImage(url, filename, subfolder) {
    let success = 0;
    await axios.get(url, { responseType: 'arraybuffer', headers: {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:98.0) Gecko/20100101 Firefox/98.0",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.5",
        "Accept-Encoding": "gzip, deflate",
        "Connection": "keep-alive",
        "Upgrade-Insecure-Requests": "1",
        "Sec-Fetch-Dest": "document",
        "Sec-Fetch-Mode": "navigate",
        "Sec-Fetch-Site": "none",
        "Sec-Fetch-User": "?1",
        "Cache-Control": "max-age=0",
        "Referer": "https://www.google.com/"
    }})
        .then(async (response) => {
            const filepath = path.join(__dirname, subfolder, filename);
            fs.writeFileSync(filepath, response.data, 'binary');
            await sleep(10000);
            success = 1;
        })
        .catch(async (error) => {
            if ((error.response) && error.response.status!=404) {
                console.log(error.code);
                console.log(error.message);
                console.log(url, subfolder);
                if (error.response) {
                    console.log(error.response.status);
                }
            }
            await sleep(5000);
        });
    return success;
}

async function scrapeMorphs() {
  const response = await axios.get(BASE_URL);
  const $ = cheerio.load(response.data);
  let currentHeader = '';

  $('.morphlist-row').each(async (index, element) => {
    const imgSrc = new URL(clean($(element).find('img').attr('src')), BASE_URL).href;
    const nameLink = $(element).find('a');
    const name = nameLink.text();
    const traits = $(element).find('.ivcombo').text().split(' + ');
    const thumbnailFilename = path.basename(imgSrc);
    let previous = $(element).prev();
    while (previous.length && !previous.hasClass('morphlist-header')) {
      previous = previous.prev();
    }
    if (previous.length) {
      currentHeader = previous.text();
    }
    if (!TYPE_MAP[currentHeader]) {
        return;
    }
    
    await downloadImage(imgSrc, thumbnailFilename, 'thumbnails');
    await sleep(5000);

    const filename = clean(nameLink.attr('href')).substring(4);
    const morphUrl = `${BASE_URL}/${filename}/pics`;
    const hatchlingImgFilename = `${filename.split("_").join('')}-a.jpg`;
    const adultImgFilename = `${filename.split("_").join('')}-h.jpg`;
    
    const hSuccess = await downloadImage(`${morphUrl}/${hatchlingImgFilename}`, hatchlingImgFilename, 'pics');
    await sleep(5000);
    const aSuccess = await downloadImage(`${morphUrl}/${adultImgFilename}`, adultImgFilename, 'pics');
    await sleep(5000);

    const jsonData = {
      name,
      type: TYPE_MAP[currentHeader],
      traits,
      thumbnail: `./thumbnails/${thumbnailFilename}`,
      hatchlingImg: hSuccess? `./pics/${hatchlingImgFilename}`: "",
      adultImg: aSuccess? `./pics/${adultImgFilename}`: ""
    };

    fs.writeFileSync(path.join(__dirname, "json", `${filename}.json`), JSON.stringify(jsonData, null, 2));
  });
}

scrapeMorphs();

File: scrape.py
import os
import json
import requests
from bs4 import BeautifulSoup, Tag

BASE_URL = "https://iansvivarium.com"
MORPHS_URL = f"{BASE_URL}/morphs/"

TYPE_MAP = {
    "Wildtypes": 0,
    "Single Recessive": 1,
    "Double Trait": 2,
    "Triple Trait": 3,
    "Quad Trait": 4,
    "Five Trait": 5,
    "Six Trait": 6,
    "Single Dominant": 7,
    "Single Incomplete Dominant": 7,
    "Selectively Bred": 8
}

def download_image(url, subdir):
    response = requests.get(url)
    if response.status_code == 200:
        with open(os.path.join(os.getcwd(), subdir), "wb") as f:
            f.write(response.content)
        return 1
    else:
        return 0

response = requests.get(MORPHS_URL)
soup = BeautifulSoup(response.text, "html.parser")
count = 0
t_count = 0
h_count = 0
a_count = 0
for li in soup.find_all("li", class_="morphlist-row"):
    current_header = None
    for sibling in li.previous_siblings:
        if isinstance(sibling, Tag) and sibling.has_attr("class") and "morphlist-header" in sibling["class"]:
            current_header = sibling.text
            break
    if (TYPE_MAP.get(current_header, -1)==-1):
        break
    
    count += 1
    
    img = li.find("img")
    thumbnail_url = f"{BASE_URL}/{img['src'].lstrip('../')}"
    thumbnail_filename = img["src"].split("/")[-1]
    t_flag = download_image(thumbnail_url, f"thumbnails/{thumbnail_filename}")
    t_count += t_flag
    anchor = li.find("a")
    morph_file_name = anchor['href'].split('?m=')[1].split('&sid')[0]
    morph_img_name = ''.join(morph_file_name.split('_'))
    morph_url = f"{BASE_URL}/morphs/{morph_file_name}/pics/"
    morph_name = anchor.text
    h_img_name = f"{morph_img_name}-h.jpg"
    a_img_name = f"{morph_img_name}-a.jpg"

    span = li.find("span", class_="ivcombo")
    traits = span.text.split(" + ")

    h_flag=download_image(f"{morph_url}{h_img_name}", f"pics/{h_img_name}")
    a_flag=download_image(f"{morph_url}{a_img_name}", f"pics/{a_img_name}")
    h_count += h_flag
    a_count += a_flag

    json_data = {
        "name": morph_name,
        "type": TYPE_MAP[current_header],
        "traits": traits,
        "thumbnail": f"./thumbnails/{thumbnail_filename}",
        "hatchlingImg": f"./pics/{h_img_name}" if h_flag else "",
        "adultImg": f"./pics/{a_img_name}" if a_flag else "",
    }

    with open(f"{os.getcwd()}/json/{morph_file_name}.json", "w") as f:
        json.dump(json_data, f, indent=2)
    b
    print(morph_file_name, t_flag, a_flag, h_flag)

print(count, t_count, a_count, h_count)
